RyuJIT: By-ref assignment with null leads to runtime crash

The following example crashes the runtime when compiled and run in either debug or release:
```csharp
class C0
{
}

struct S0
{
    public C0 F0;
    public ulong F4;
}

class C1
{
    public S0 F3;
}

struct S1
{
    public S0 F3;
}

public class Program
{
    static S1 s_38;
    static C1 s_43;

    public static void Main()
    {
        s_38.F3 = s_43.F3;
    }
}
```

In debug (and probably checked) builds, this assertion triggers:
```
Assert failure(PID 20848 [0x00005170], Thread: 17420 [0x440c]): Consistency check failed: AV in clr at this callstack:
------
CORECLR! JIT_ByRefWriteBarrier + 0x0 (0x00007ffc`5e207d70)
<no module>! <no symbol> + 0x0 (0x00007ffb`fef22611)
<no module>! <no symbol> + 0x0 (0x00007ffb`fee045e0)
<no module>! <no symbol> + 0x0 (0x000000fa`00000005)
<no module>! <no symbol> + 0x0 (0x000000fa`2077d580)
-----
.AV on tid=0x440c (17420), cxr=000000FA2077CE00, exr=000000FA2077D2F0
FAILED: false

CORECLR! CHECK::Trigger + 0x275 (0x00007ffc`5db0cea5)
CORECLR! CLRVectoredExceptionHandlerPhase3 + 0x332 (0x00007ffc`5dc1bd62)
CORECLR! CLRVectoredExceptionHandlerPhase2 + 0x8C (0x00007ffc`5dc1b70c)
CORECLR! CLRVectoredExceptionHandler + 0x275 (0x00007ffc`5dc1b665)
CORECLR! CLRVectoredExceptionHandlerShim + 0x18D (0x00007ffc`5dc1c04d)
NTDLL! RtlInitializeCriticalSection + 0x1D8 (0x00007ffc`c5bd5678)
NTDLL! RtlWalkFrameChain + 0x109A (0x00007ffc`c5b7692a)
NTDLL! KiUserExceptionDispatcher + 0x2E (0x00007ffc`c5c0dc1e)
CORECLR! JIT_ByRefWriteBarrier + 0x0 (0x00007ffc`5e207d70)
<no module>! <no symbol> + 0x0 (0x00007ffb`fef22611)
    File: z:\programming\dotnet\coreclr\src\vm\excep.cpp Line: 7831
    Image: Z:\Programming\dotnet\coreclr\bin\Product\Windows_NT.x64.Debug\CoreRun.exe
```

The code generated is:
```asm
; Assembly listing for method Program:Main()
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 OutArgs      [V00    ] (  1,  1   )  lclBlk (32) [rsp+0x00]
;  V01 cse0         [V01,T00] (  5,  5   )    long  ->  [rsp+0x20]
;
; Lcl frame size = 40

G_M5092_IG01:
       57                   push     rdi
       56                   push     rsi
       4883EC28             sub      rsp, 40

G_M5092_IG02:
       48B9E045E3FEFB7F0000 mov      rcx, 0x7FFBFEE345E0
       BA05000000           mov      edx, 5
       E8461E455F           call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       48B87829001091010000 mov      rax, 0x19110002978
       488B00               mov      rax, gword ptr [rax]
       488D7008             lea      rsi, bword ptr [rax+8]
       48B87029001091010000 mov      rax, 0x19110002970
       488B00               mov      rax, gword ptr [rax]
       488D7808             lea      rdi, bword ptr [rax+8]
       E85F572B5F           call     CORINFO_HELP_ASSIGN_BYREF
       48A5                 movsq

G_M5092_IG03:
       4883C428             add      rsp, 40
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret

; Total bytes of code 74, prolog size 6 for method Program:Main()
;
```

Question: How will this normally be handled by CoreCLR? Should the runtime turn the AV in `CORINFO_HELP_ASSIGN_BYREF` into a managed NRE, or is the caller supposed to null-check/deref the reference before calling `CORINFO_HELP_ASSIGN_BYREF`?

I tried making the simple change to add `JIT_ByRefWriteBarrier` to `IsIPInMarkedJitHelper`, but I get an error in the build of dactablegen:
```
CUSTOMBUILD : warning CS1668: Invalid search path 'lib\um\x64' specified in 'LI
B environment variable' -- 'directory does not exist' [C:\git\coreclr\bin\obj\W
indows_NT.x64.Checked\src\ToolBox\SOS\DacTableGen\dactablegen.vcxproj]
c:\git\coreclr\src\vm\excep.cpp(6980): error C2065: 'JIT_ByRefWriteBarrier_End'
: undeclared identifier [C:\git\coreclr\bin\obj\Windows_NT.x64.Checked\src\vm\w
ks\cee_wks.vcxproj]
```
Before I spend a lot of time tracking this down, can someone point me in the right direction? @jkotas @janvorli

--- a/src/vm/amd64/JitHelpers_Fast.asm
+++ b/src/vm/amd64/JitHelpers_Fast.asm
@@ -722,7 +722,7 @@ endif
         add     rdi, 8h
         add     rsi, 8h
         ret
-LEAF_END JIT_ByRefWriteBarrier, _TEXT
+LEAF_END_MARKED JIT_ByRefWriteBarrier, _TEXT
 
 
 g_pObjectClass      equ     ?g_pObjectClass@@3PEAVMethodTable@@EA

--- a/src/vm/amd64/jithelpers_fast.S
+++ b/src/vm/amd64/jithelpers_fast.S
@@ -384,7 +384,7 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
         add     rdi, 8h
         add     rsi, 8h
         ret
-LEAF_END JIT_ByRefWriteBarrier, _TEXT
+LEAF_END_MARKED JIT_ByRefWriteBarrier, _TEXT
 
 // TODO: put definition for this in asmconstants.h
 #define CanCast 1

--- a/src/vm/excep.cpp
+++ b/src/vm/excep.cpp
@@ -6951,6 +6951,7 @@ EXTERN_C void JIT_MemCpy_End();
 
 EXTERN_C void JIT_WriteBarrier_End();
 EXTERN_C void JIT_CheckedWriteBarrier_End();
+EXTERN_C void JIT_ByRefWriteBarrier_End();
 #endif // _TARGET_X86_
 
 #if defined(_TARGET_AMD64_) && defined(_DEBUG)
@@ -6977,6 +6978,7 @@ bool IsIPInMarkedJitHelper(UINT_PTR uControlPc)
 
     CHECK_RANGE(JIT_WriteBarrier)
     CHECK_RANGE(JIT_CheckedWriteBarrier)
+    CHECK_RANGE(JIT_ByRefWriteBarrier)
 #else
 #ifdef FEATURE_PAL
     CHECK_RANGE(JIT_WriteBarrierGroup)

--- a/tests/src/JIT/Regression/JitBlue/GitHub_19444/GitHub_19444.cs
+++ b/tests/src/JIT/Regression/JitBlue/GitHub_19444/GitHub_19444.cs
@@ -1,0 +1,69 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// This test case illustrates a bug where the JIT_ByRefWriteBarrier was not
+// included in IsIPInMarkedJitHelper on non-32-bit-x86 platforms.
+
+using System;
+
+class C0
+{
+}
+
+struct S0
+{
+    public C0 F0;
+    public ulong F4;
+}
+
+class C1
+{
+    public S0 F3;
+}
+
+struct S1
+{
+    public S0 F3;
+}
+
+struct S2
+{
+    public C0 F0;
+    public uint F4;
+}
+
+class C3
+{
+    public S2 F3;
+}
+
+struct S3
+{
+    public S2 F3;
+}
+
+public class GitHub_19444
+{
+    static S1 s_38;
+    static C1 s_43;
+    static S3 s_1;
+    static C3 s_2;
+
+    public static int Main()
+    {
+        try
+        {
+            s_38.F3 = s_43.F3;
+            s_1.F3 = s_2.F3;
+        }
+        catch (System.NullReferenceException)
+        {
+            Console.WriteLine("PASS");
+            return 100;
+        }
+        Console.WriteLine("FAIL");
+        return -1;
+    }
+}
+

--- a/tests/src/JIT/Regression/JitBlue/GitHub_19444/GitHub_19444.csproj
+++ b/tests/src/JIT/Regression/JitBlue/GitHub_19444/GitHub_19444.csproj
@@ -1,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <AssemblyName>$(MSBuildProjectName)</AssemblyName>
+    <SchemaVersion>2.0</SchemaVersion>
+    <ProjectGuid>{95DFC527-4DC1-495E-97D7-E94EE1F7140D}</ProjectGuid>
+    <OutputType>Exe</OutputType>
+    <ProjectTypeGuids>{786C830F-07A1-408B-BD7F-6EE04809D6DB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
+  </PropertyGroup>
+  <!-- Default configurations to help VS understand the configurations -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' "></PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' "></PropertyGroup>
+  <ItemGroup>
+    <CodeAnalysisDependentAssemblyPaths Condition=" '$(VS100COMNTOOLS)' != '' " Include="$(VS100COMNTOOLS)..\IDE\PrivateAssemblies">
+      <Visible>False</Visible>
+    </CodeAnalysisDependentAssemblyPaths>
+  </ItemGroup>
+  <PropertyGroup>
+    <DebugType></DebugType>
+    <Optimize>True</Optimize>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="$(MSBuildProjectName).cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
+  </ItemGroup>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
+  <PropertyGroup Condition=" '$(MsBuildProjectDirOverride)' != '' "></PropertyGroup>
+</Project>

